src/error.jl

```diff
@@ -2,10 +2,63 @@ struct MPIError <: Exception
     code::Cint
 end
 
-macro mpichk(expr)
+
+"""
+    FeatureLevelError
+
+This error is thrown if a feature is not implemented in the current MPI backend.
+"""
+struct FeatureLevelError <: Exception
+    function_name::String
+    min_feature_level::VersionNumber # minimal MPI version required for this feature to be available
+end
+
+function Base.show(io::IO, err::FeatureLevelError)
+    print(io, "FeatureLevelError($(err.function_name)): Minimum version is $(err.min_feature_level)")
+end
+
+macro mpichk(expr, min_ver_expr=v"1.0")
+    # This macro can currently just handle ccall expressions.
+    @assert expr isa Expr && expr.head == :call && expr.args[1] == :ccall
+    
+    # A ccall can have either a tuple as its first argument or just the symbol, 
+    # if the called function does not have input arguments.
+    if expr.args[2].head == :tuple
+        mpi_method_name_meta = expr.args[2].args[1]
+    elseif expr.args[2].head == :symbol
+        mpi_method_name_meta = expr.args[2]
+    end
+
+    # Adjust MPI api call to platform-specific details.  
     expr = macroexpand(@__MODULE__, :(@mpicall($expr)))
-    # MPI_SUCCESS is defined to be 0
-    :((errcode = $(esc(expr))) == 0 || throw(MPIError(errcode)))
+    
+    # Generate the check code
+    quote
+        # Check if MPI library could be found
+        mpihandle = Libc.Libdl.dlopen(libmpi, throw_error=false)
+        if mpihandle == nothing
+            error("Could not open MPI library '$(libmpi)'")
+        end
+
+        # Interpolate symbol from macro into code
+        mpi_method_symbol = $(esc(mpi_method_name_meta))
+
+        # If the symbol can be found in the library we make the actual call to the MPI function
+        if Libc.Libdl.dlsym(mpihandle, mpi_method_symbol, throw_error=false) != nothing
```

Comment:
This looks like an unacceptably hefty runtime check:
```
julia> @benchmark Libdl.dlsym(Libdl.dlopen("libmpi.12.dylib"), :MPI_Send)
BenchmarkTools.Trial: 10000 samples with 1 evaluation.
 Range (min … max):  11.208 μs … 68.875 μs  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     11.833 μs              ┊ GC (median):    0.00%
 Time  (mean ± σ):   12.057 μs ±  1.506 μs  ┊ GC (mean ± σ):  0.00% ± 0.00%

      ▂█▄▂▃
  ▂▄███████▅▄▃▂▂▂▁▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▂
  11.2 μs         Histogram: frequency by time          17 μs <

 Memory estimate: 0 bytes, allocs estimate: 0.
```
This is a latency benchmark of `Send`-`Recv!` (ping-pong) on Fugaku with a recent `master` of `MPI.jl`: https://github.com/giordano/julia-on-fugaku/blob/372f058d8b899bddc18cefc6326ea2fb2cd8034f/benchmarks/mpi-point-to-point/pingpong-latency.pdf  With this change, latency at sizes up to ~32 KiB would be completely dwarfed by this single check, paying this price at every single call (and on Fugaku I'd expect `dlopen`/`dlsym` to be much slower than on my laptop, making things much worse).  If anything, this should be a compile-time check.

Reply:
what is the latency of a `try`/`catch`?

Reply:
If this is a reliable test:
```
julia> @benchmark f()
BenchmarkTools.Trial: 10000 samples with 1000 evaluations.
 Range (min … max):  0.875 ns … 7.042 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     0.958 ns             ┊ GC (median):    0.00%
 Time  (mean ± σ):   0.954 ns ± 0.063 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                                                          █
  ▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▃▃▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁█ ▂
  0.875 ns       Histogram: frequency by time      0.959 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.

julia> @benchmark g()
BenchmarkTools.Trial: 10000 samples with 997 evaluations.
 Range (min … max):  20.060 ns … 39.451 ns  ┊ GC (min … max): 0.00% … 0.00%
 Time  (median):     20.352 ns              ┊ GC (median):    0.00%
 Time  (mean ± σ):   20.359 ns ±  0.295 ns  ┊ GC (mean ± σ):  0.00% ± 0.00%

                     ▃    ▆    ▇    █    ▄    ▁    ▂
  ▂▁▁▁▂▁▁▁▁▃▁▁▁▁▆▁▁▁▁█▁▁▁▁█▁▁▁▁█▁▁▁▁█▁▁▁▁█▁▁▁▁█▁▁▁▁█▁▁▁▁▅▁▁▁▂ ▃
  20.1 ns         Histogram: frequency by time        20.6 ns <

 Memory estimate: 0 bytes, allocs estimate: 0.
```

But again, I think any test like this should be at-compile time (with `@static` or a top-level `if§): the symbol would either be available or not for the currently used MPI library, we don't have to pay the price for a check at runtime.

Reply:
this is getting massively out of scope for this PR though

Reply:
We do quite a bit of work during startup setting various MPI constants. It would be straightforward to set a global variable there indicating whether this function is available, and to check that variable here.

Reply:
I would prefer to make the change of the error management a separater PR, since this really takes more time than it should.

Reply:
You need to look for the function at run time, not at compile time. You need to schedule it the same way as done e.g. in `consts.jl`.

```Julia
const stuff_works = Ref{Bool}(false)
function go()
    stuff_works[] = (... check whether stuff is working ...)
end
push!(initexprs, go)
```
Place this code near the end of the file `consts.jl`.

(Choose better function/variable names.)

Reply:
> I would prefer to make the change of the error management a separater PR, since this really takes more time than it should.

Agreed!